detector/calorimeter/SteppedMuonBarrel_o2_v02_geo.cpp

```diff
@@ -0,0 +1,280 @@
+
+    DetElement sdet(det_name, x_det.id());
+    //DetElement stave("stave0", x_det.id());
+    //Volume motherVol = theDetector.pickMotherVolume(sdet);
+    Readout readout = sens.readout();
+    Segmentation seg = readout.segmentation();
+    
+    std::vector<double> cellSizeVector = seg.segmentation()->cellDimensions(0); //Assume uniform cell sizes, provide dummy cellID
+    double cell_sizeX = cellSizeVector[0];
+    double cell_sizeY = cellSizeVector[1];
+    
+    //Create caloData object to extend driver with data required for reconstruction
+    LayeredCalorimeterData* caloData = new LayeredCalorimeterData ;
+    caloData->layoutType = LayeredCalorimeterData::BarrelLayout ;
+    caloData->inner_symmetry = nsides;
+    caloData->outer_symmetry = nsides; 
+    
+    /*  NOTE: phi0=0 means lower face flat parallel to experimental floor
+     *  This is achieved by rotating the modules with respect to the envelope
+     *  which is assumed to be a Polyhedron and has its axes rotated with respect
+     *  to the world by 180/nsides. In any other case (e.g. if you want to have
+     *  a tip of the calorimeter touching the ground) this value needs to be computed 
+     */
+    
+    caloData->inner_phi0 = 0.; 
+    caloData->outer_phi0 = 0.; 
+    caloData->gap0 = 0.; //FIXME
+    caloData->gap1 = 0.; //FIXME
+    caloData->gap2 = 0.; //FIXME  
+    
+    for (xml_coll_t c(x_det, _U(layer)); c; ++c) {
+        xml_comp_t x_layer = c;
+        int repeat = x_layer.repeat();
+        totalRepeat += repeat;
+        totalSlices += x_layer.numChildren(_U(slice));
+    }
+    
+    // CHECK THIS!
+    // --- create an envelope volume and position it into the world ---------------------
+    
+    Volume envelopeVol = dd4hep::xml::createPlacedEnvelope( theDetector,  e , sdet ) ;
+    dd4hep::xml::setDetectorTypeFlag( e, sdet ) ;
+    
+    if( theDetector.buildType() == BUILD_ENVELOPE ) return sdet ;
+    
+    //-----------------------------------------------------------------------------------
+    
+    // Extent of the calorimeter in the r-z-plane [ rmin, rmax, zmin, zmax ] in mm.
+    caloData->extent[0] = rmin ;
+    caloData->extent[1] = rmin + totalThickness;
+    caloData->extent[2] = 0. ;
+
+    double innerAngle = 2 * M_PI / nsides;
+    double tan_inner = std::tan(innerAngle/2.) * 2;
+    double layer_pos_z = -(totalThickness / 2);
+    double layer_dim_x = rmin * tan_inner/2. - gap * 2;
+  
+    int layer_num = 0;
+    int layerType   = 0;
+
+    for (xml_coll_t c(x_det, _U(layer)); c; ++c) {
+        xml_comp_t x_layer = c;
+        int repeat = x_layer.repeat();            // Get number of times to repeat this layer.
+        const Layer* lay = layering.layer(layer_num); // Get the layer from the layering engine.
+        double layer_thickness = lay->thickness();
+        string layer_type_name = _toString(layerType, "layerType%d");
+
+	// This needs checked!
+	caloData->extent[3] = detZ/2.0 + floor(repeat/2)*totalThickness/repeat;
```

Comment:
Is this correct? Frank or Andre, could you please have a look?

Reply:
What is this supposed to do?

Reply:
In the previous version, caloData extent in z was defined as dz of the stave. However, here there's no stave as we have a stepped geometry. So the question is what should extent[3] be: the dz of the layer or the dz of the longest layer?

Reply:
There might be assumptions in the reconstruction that assume Z_hit > Z_Max_Barrel means endcap hit, so things will fail in either case at some point.

Reply:
Indeed, that is what I am worried about. Setting it the other way round might just transfer this problem to the endcap.